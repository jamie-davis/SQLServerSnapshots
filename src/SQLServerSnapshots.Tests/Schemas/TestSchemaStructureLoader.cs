using System.ComponentModel;
using SQLServerSnapshots.Schemas;
using SQLServerSnapshots.Tests.Configuration;
using TestConsoleLib;
using TestConsoleLib.Testing;
using Xunit;

namespace SQLServerSnapshots.Tests.Schemas
{
    [Category("localhost Dependent")]
    public class TestSchemaStructureLoader
    {
        private const string CreateTables =
            @"
CREATE SCHEMA Test;
GO
CREATE TABLE [Test].[A]
(
    Id INT PRIMARY KEY IDENTITY,
    Name VARCHAR(100) NULL,
    CreatedDate DATETIME NOT NULL DEFAULT(GETDATE())
);
GO
";

        private const string CreateRelatedTables =
            @"
CREATE SCHEMA Test;
GO
CREATE TABLE [Test].[A_Main]
(
    MainId INT PRIMARY KEY IDENTITY,
    Name VARCHAR(100) NULL,
    CreatedDate DATETIME NOT NULL DEFAULT(GETDATE())
);
GO
CREATE TABLE [Test].[B_Related]
(
    BId INT PRIMARY KEY IDENTITY,
    Name VARCHAR(100) NULL,
    MainId INT NOT NULL,
CONSTRAINT fk_B_MainId
    FOREIGN KEY (MainId)
    REFERENCES [Test].[A_Main] (MainId)
);
GO
CREATE TABLE [C_Non_Schema]
(
    CId INT PRIMARY KEY IDENTITY,
    Name VARCHAR(100) NULL,
    MainId INT NOT NULL,
CONSTRAINT fk_B_MainId
    FOREIGN KEY (MainId)
    REFERENCES [Test].[A_Main] (MainId)
);
GO
";


        public TestSchemaStructureLoader()
        {
            DbController.TearDown();
        }

        [Fact]
        public void BasicTableStructureIsLoaded()
        {
            //Arrange
            DbController.Run(CreateTables);

            //Act
            var result = SchemaStructureLoader.Load(DbController.Server, DbController.TestDbName, "Test");

            //Assert
            var output = new Output();
            result.Report(output);
            output.Report.Verify();
        }

        [Fact]
        public void RelationshipsAreExtracted()
        {
            //Arrange
            DbController.Run(CreateRelatedTables);

            //Act
            var result = SchemaStructureLoader.Load(DbController.Server, DbController.TestDbName, "Test");

            //Assert
            var output = new Output();
            result.Report(output);
            output.Report.Verify();
        }
    }
}
