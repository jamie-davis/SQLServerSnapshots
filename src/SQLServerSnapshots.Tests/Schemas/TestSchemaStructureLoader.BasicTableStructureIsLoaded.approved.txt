[SqlServerSnapshots.Tests].[Test]                                                              
---------------------------------                                                              
Table Name                                                                                     
----------                                                                                     
[Test].[A]                                                                                     
                                                                                               
                In                                                                             
                Primary                                                                        
    Name        Key                                                                            
    ----------- -------                                                                        
    Id          True                                                                           
    Name        False                                                                          
    CreatedDate False                                                                          
                                                                                               
    Referencing       Referenced                                                               
    Columns     Table Columns                                                                  
    ----------- ----- ----------                                                               
                                                                                               
                                                                                               